File: Java-version/Turtle.java
import java.awt.*;
import java.util.Random;
import java.util.RandomAccess;

public class Turtle extends Animal{

    public Turtle(int x, int y, World world)
    {super(x,y,world);
        sign = 'T';
        name = "Turtle";
        color = Color.BLUE;
        str = 2;
        ini = 1;
        speciesId = 4;
    }
    public Turtle(int str, int ini, int xPos, int yPos, int age, int oId, boolean moved, World world)
    {super(str, ini, xPos, yPos, age, oId, moved, world);
        sign = 'T';
        name = "Turtle";
        speciesId = 4;
        color = Color.BLUE;
    }
    @Override
    public Animal MakeNewA(int x, int y) {
        Animal newAnimal = new Turtle(x, y, world);
        return newAnimal;
    }
    @Override
    public void TakeAction() {
        if (age != DEAD && !moved)
        {
            int changeX = 0, changeY = 0, rNum;
            Random rand = new Random();
            rNum = rand.nextInt(100);
            if (rNum < 25)
            {
                while (true)
                {
                    rNum = rand.nextInt(4);
                    if (rNum == 0) {
                        changeX = -1;
                    }
                    else if (rNum == 1) {
                        changeX = 1;
                    }
                    else if (rNum == 2) {
                        changeY = -1;
                    }
                    else {
                        changeY = 1;
                    }
                    if (xPos + changeX < world.GetW() && yPos + changeY < world.GetH() &&  xPos + changeX >= 0 && yPos + changeY >= 0)
                        break;
                    else {
                        changeX = 0;
                        changeY = 0;
                    }
                }

                Organism potentialCollision = world.LookForCollision(xPos + changeX, yPos + changeY);
                if (potentialCollision != null) {
                    Collision(potentialCollision);
                }
                if (age != DEAD)
                {
                    potentialCollision = world.LookForCollision(xPos + changeX, yPos + changeY); // recheck if collision changed the state of disputed tile
                    if (potentialCollision == null)
                    {
                        int tmpx = xPos + changeX;
                        int tmpy = yPos + changeY;
                        System.out.print( name + " moved from tile (" + xPos + ", " + yPos + ") to tile (" + tmpx + ", " + tmpy + ")" + '\n');
                        log = log + name + " moved from tile (" + xPos + ", " + yPos + ") to tile (" + tmpx + ", " + tmpy + ")" + '\n';
                        xPos += changeX;
                        yPos += changeY;
                    }
                }
            }
            else {
                System.out.print(name + " from tile (" + xPos + ", " + yPos + ") stayed in place" + '\n');
                log = log + name + " from tile (" + xPos + ", " + yPos + ") stayed in place" + '\n';
            }
            moved = true;
        }
        world.AddLog(log);
        log = " ";
        if (age > 0) age++;
    }
    @Override
    public void Win(Organism otherCreature)
    {
        if (deflected) {
            System.out.print( " and " + name + " won! " + otherCreature.GetName() + " was pushed back" + '\n');
            log = log + " and " + name + " won! " + otherCreature.GetName() + " was pushed back" + '\n';
            deflected = false;
        }
        else {
            System.out.print( " and " + name + " won! " + otherCreature.GetName() + " was eaten" + '\n');
            log = log + " and " + name + " won! " + otherCreature.GetName() + " was pushed back" + '\n';
            world.RmOrganism(otherCreature.GetOId());
            world.AddLog(otherCreature.GetLog());
            otherCreature.ResetLog();
            otherCreature.SetAge(DEAD);
        }
    }
    @Override
    public boolean Defended(Organism attackingCreature) {
        if (attackingCreature.GetS() < 5) {
            deflected = true;
            return true;
        }
        return false;
    }
    private boolean deflected;
}

File: Python-version/turtle.py
from animal import Animal
import random

class Turtle(Animal):
    def __init__(self, x, y, world, oid=None, s=None, i=None, a=None, m=None):
        super().__init__(x, y, world, oid, str, i, a, m)
        self.sign = 'T'
        self.name = "Turtle"
        self.color = "blue"
        self.str = 2
        self.ini = 1
        self.speciesId = 4
        self.deflected=False

    def MakeNewA(self, x, y):
        return Turtle(x, y, self.world)

    def TakeAction(self):
        if self.age != -1 and not self.moved:
            changex=0
            changey=0
            rand=random.Random()
            rnum=rand.randint(0,99)
            if rnum<25:

                while(True):
                    rnum=rand.randint(0, 3)
                    if rnum == 0:
                        changex=-1
                    elif rnum == 1:
                        changex=1
                    elif rnum == 2:
                        changey=-1
                    elif rnum == 3:
                        changey=1
                    if self.xPos+changex<self.world.GetW() and self.yPos+changey<self.world.GetH() and self.xPos+ changex>=0 and self.yPos+changey>=0:
                        break
                    else:
                        changey=0
                        changex=0


                potential_collision = self.world.LookForCollision(self.xPos + changex, self.yPos + changey)

                if potential_collision is not None:
                    self.Collision(potential_collision)

                if self.age != -1:
                    potential_collision = self.world.LookForCollision(self.xPos + changex, self.yPos + changey)
                    if potential_collision is None:
                        tmpx = self.xPos + changex
                        tmpy = self.yPos + changey
                        print(f"{self.name} moved from tile ({self.xPos}, {self.yPos}) to tile ({tmpx}, {tmpy})")
                        self.log=f"{self.name} moved from tile ({self.xPos}, {self.yPos}) to tile ({tmpx}, {tmpy})"
                        self.xPos+=changex
                        self.yPos+=changey

            else:
                print(f" {self.name} from tile ({self.xPos}, {self.yPos}) stayed in place")

        self.moved=True
        self.world.AddLog(self.log)
        self.log=" "
        if self.age>0:
            self.age+=1

    def Win(self,othercreature):
        if self.deflected:
            print(f" and {self.name} won! {othercreature.GetName()} was pushed back")
            self.deflected=False
        else:
            print(f" and {self.name} won! {othercreature.GetName()} was eaten")
            self.world.RmOrganism(othercreature.GetOId())
            self.world.AddLog(othercreature.GetLog())
            othercreature.ResetLog()
            othercreature.SetAge(-1)

    def Defended(self,attackingcreature):
        if attackingcreature.GetS()<5:
            self.deflected=True
            return True
        return False
